### PR TITLE
RISC-V: Support ftrace for core and TA

### PR DESCRIPTION
**Pull request by gagachang**
This PR aims to support ftrace for RISC-V architecture.
Note that this PR may depends on WIP PR https://github.com/OP-TEE/optee_os/pull/6104.
So this PR must be merged only after the mentioned PR is merged.

To test ftrace on RISC-V architecture using QEMU:
1. Enable these ftrace configurations from `core/arch/riscv/plat-virt/conf.mk`
```bash
$(call force,CFG_FTRACE_SUPPORT,y)
$(call force,CFG_SYSCALL_FTRACE,y)
$(call force,CFG_ULIBS_MCOUNT,y)
$(call force,CFG_TA_MCOUNT,y)
```
2. Manually enlarge the TA ftrace buffer from `ta/arch/riscv/ta.ld.S`:
```C
#ifndef CFG_FTRACE_BUF_SIZE
#define CFG_FTRACE_BUF_SIZE 65536
#endif
```
3. Enable QEMU virtfs auto mount. (QEMU_VIRTFS_AUTOMOUNT=y)
4. Start QEMU, run `xtest regression_1004`, and copy out the ftrace result to host:
```
Welcome to Buildroot, type root or test to login
buildroot login: root
# xtest regression_1004
...
+-----------------------------------------------------
Result of testsuite regression filtered by "regression_1004":
regression_1004 OK
+-----------------------------------------------------
10 subtests of which 0 failed
1 test case of which 0 failed
103 test cases were skipped
TEE test application done!
# cp /tmp/ftrace-cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.out /mnt/host/
# QEMU: Terminated
```
5. Run `scripts/symbolize.py` to see the result:
```
$ cat ftrace-cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.out | optee_os/scripts/symbolize.py \
         -d optee_os/out/riscv/core \
         -d out-br/build/optee_test_ext-1.0/ta/*/out | less
TEE load address @ 0x8e000000
Function graph for TA: cb3e5ba0-adf1-11e0-998b-0002a5d5c51b @ 40060000
              | __ta_entry() {
              |  __utee_entry() {
              |   entry_open_session() {
              |    ta_header_add_session() {
   138.500 us |     ta_header_get_session();
              |     init_instance() {
     7.500 us |      tahead_get_trace_level();
    29.300 us |      trace_set_level();
     6.700 us |      __utee_gprof_init();
              |      malloc_add_pool() {
              |       gen_malloc_add_pool() {
     7.400 us |        malloc_lock();
              |        raw_malloc_add_pool() {
     7.500 us |         tag_asan_free();
    61.600 us |         bpool();
              |         realloc_unlocked() {
              |          raw_realloc() {
     7.200 us |           raw_malloc_validate_pools();
              |           bget() {
    49.300 us |            buf_get_pos();
   388.500 us |           }
              |           raw_malloc_return_hook() {
    13.400 us |            maybe_tag_buf();
    55.000 us |           }
   593.900 us |          }
   620.400 us |         }
   833.100 us |        }
     7.400 us |        malloc_unlock();
   910.100 us |       }
   936.900 us |      }
              |      get_memtag_implementation() {
              |       TEE_GetPropertyAsU32() {
    33.200 us |        is_propset_pseudo_handle();
              |        __utee_check_instring_annotation() {
    75.700 us |         strlen();
              |         check_access() {
              |          TEE_CheckMemoryAccessRights() {
              |           tee_obj_attr_copy_from() {
              |            __mutex_read_unlock() {
   138.800 us |             __mutex_read_unlock();
   461.900 us |            }
    14.700 us |            op_attr_secret_value_to_user();
    13.200 us |            op_attr_secret_value_from_user();
              |            core_init_mmu_map() {
              |             assign_mem_va() {
              |              new_session_id() {
     3.700 us |               new_session_id();

......
```

### CODE REVIEW

File: core/arch/riscv/include/riscv.h [hunk 1]
```diff
@@ -374,6 +374,19 @@ static inline __noprof void uret(void)
 	asm volatile("uret");
 }
 
+__noprof uint64_t read_time(void);
+
+static inline __noprof uint64_t barrier_read_counter_timer(void)
+{
+	mb();
```
Comment (jforissier) [inline core/arch/riscv/include/riscv.h:381]: Maybe add
```C
	/* We want the timer value after pending operations have completed */
```
...to silence the checkpatch warning.

**Review by jforissier — COMMENTED** [review]
Hello @gagachang,

This series LGTM with one minor comment below.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Regarding the instructions for testing, please note a single command is enough, you do not need to edit the platform configuration files etc.:
```
make CFG_FTRACE_SUPPORT=y CFG_SYSCALL_FTRACE=y CFG_ULIBS_MCOUNT=y CFG_TA_MCOUNT=y CFG_FTRACE_BUF_SIZE=65536 QEMU_VIRTFS_AUTOMOUNT=y run
```

Please also note https://github.com/OP-TEE/optee_os/pull/6119 which is mostly ready for merging and should hopefully not break anything, but you may want to test with it to make sure.

File: core/arch/riscv/include/riscv.h (hunk 1 above)
Comment (gagachang) [reply core/arch/riscv/include/riscv.h]: Thanks! I've added the comments for those barriers.

**Comment by gagachang** [comment]
> Hello @gagachang,
> 
> This series LGTM with one minor comment below.
> 
> `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Thanks for review! I'll add the tag later.

> Regarding the instructions for testing, please note a single command is enough, you do not need to edit the platform configuration files etc.:
> 
> ```
> make CFG_FTRACE_SUPPORT=y CFG_SYSCALL_FTRACE=y CFG_ULIBS_MCOUNT=y CFG_TA_MCOUNT=y CFG_FTRACE_BUF_SIZE=65536 QEMU_VIRTFS_AUTOMOUNT=y run
> ```

Good. :)) I try this way and it also works well, 

> Please also note #6119 which is mostly ready for merging and should hopefully not break anything, but you may want to test with it to make sure.

Will try it later.

**Comment by gagachang** [comment]
Hello @maroueneboubakri 

Would you also check this PR if you have time?
Some changed files are submitted by you before, your review is appreciated!

**Comment by gagachang** [comment]
> Please also note #6119 which is mostly ready for merging and should hopefully not break anything, but you may want to test with it to make sure.

I test it with the same xtest regression_1004.
Seems everything works well:

```
$ optee_os/scripts/ftrace_format.py \
    ftrace-cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.out |
    optee_os/scripts/symbolize.py \
    -d optee_os/out/riscv/core \
    -d out-br/build/optee_test_ext-1.0/ta/*/out | less

TEE load address @ 0x8e000000
Function graph for TA: cb3e5ba0-adf1-11e0-998b-0002a5d5c51b @ 4004a000
            |   1 |   __ta_entry() {
            |   2 |     __utee_entry() {
            |   3 |       entry_open_session() {
            |   4 |         ta_header_add_session() {
  46.000 us |   5 |           ta_header_get_session()
            |   5 |           init_instance() {
   7.700 us |   6 |             tahead_get_trace_level()
  27.500 us |   6 |             trace_set_level()
   7.200 us |   6 |             __utee_gprof_init()
            |   6 |             malloc_add_pool() {
            |   7 |               gen_malloc_add_pool() {
   8.800 us |   8 |                 malloc_lock()
            |   8 |                 raw_malloc_add_pool() {
   7.100 us |   9 |                   tag_asan_free()
  69.800 us |   9 |                   bpool()
            |   9 |                   realloc_unlocked() {
            |  10 |                     raw_realloc() {
   7.600 us |  11 |                       raw_malloc_validate_pools()
            |  11 |                       bget() {
  52.700 us |  12 |                         buf_get_pos()
 303.100 us |  11 |                       }
            |  11 |                       raw_malloc_return_hook() {
  13.400 us |  12 |                         maybe_tag_buf()
  55.000 us |  11 |                       }
 495.100 us |  10 |                     }
 522.300 us |   9 |                   }
 741.500 us |   8 |                 }
   7.200 us |   8 |                 malloc_unlock()
 821.900 us |   7 |               }
 846.200 us |   6 |             }
            |   6 |             get_memtag_implementation() {
            |   7 |               TEE_GetPropertyAsU32() {
  34.300 us |   8 |                 is_propset_pseudo_handle()
            |   8 |                 __utee_check_instring_annotation() {
  78.500 us |   9 |                   strlen()
            |   9 |                   check_access() {
            |  10 |                     TEE_CheckMemoryAccessRights() {
            |  11 |                       syscall_check_access_rights() {
            |  12 |                         ts_get_current_session() {
  24.000 us |  13 |                           ts_get_current_session_may_fail()
 125.900 us |  12 |                         }
  13.800 us |  12 |                         to_user_ta_ctx()
  13.300 us |  12 |                         memtag_strip_tag_vaddr()
            |  12 |                         vm_check_access_rights() {
            |  13 |                           vm_buf_is_inside_um_private() {
            |  14 |                             core_is_buffer_inside() {
   3.700 us |  15 |                               is_valid_conf_and_notnull_size()
```

File: lib/libutils/ext/arch/riscv/mcount_rv.S [hunk 2]
```diff
@@ -0,0 +1,127 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2023 Andes Technology Corporation
+ * Copyright (c) 2016, Linaro Limited
+ */
+
+#include <asm.S>
+
+#if defined(CFG_TA_GPROF_SUPPORT) || defined(CFG_FTRACE_SUPPORT)
```
Comment (b49020) [inline lib/libutils/ext/arch/riscv/mcount_rv.S:9]: Does this patch-set enables support for gprof too on RISC-V? If not we should include corresponding `CFG_` checks.

**Comment by maroueneboubakri** [comment]
LGTM !

`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

Thanks
Maro

File: lib/libutils/ext/arch/riscv/mcount_rv.S (hunk 2 above)
Comment (gagachang) [reply lib/libutils/ext/arch/riscv/mcount_rv.S:9]: @b49020 Gprof support is future work.
I will remove related configurations first and add them back once RISC-V supports it.
Thanks for recommendation!

Comment (gagachang) [reply lib/libutils/ext/arch/riscv/mcount_rv.S]: Hi @b49020 

Commit https://github.com/OP-TEE/optee_os/pull/6129/commits/32552034144eb0d79cbd0142168cccb2a46723b6 removes `CFG_TA_GPROF_SUPPORT`.
Please check it.

**Comment by b49020** [comment]
`Acked-by: Sumit Garg <sumit.garg@linaro.org>`

**Comment by gagachang** [comment]
@maroueneboubakri @b49020 
Thank you for review!
Tags are applied.

Hello @jforissier 
May I know how to fix IBART fail? I have no idea about that.

**Comment by jforissier** [comment]
> @maroueneboubakri @b49020 Thank you for review! Tags are applied.
> 
> Hello @jforissier May I know how to fix IBART fail? I have no idea about that.

It looks like a transient network error. I have restarted the job. I believe anyone with a Google account can restart a job, but I don't mind doing it when needed.

**Comment by gagachang** [comment]
> It looks like a transient network error. I have restarted the job. I believe anyone with a Google account can restart a job, but I don't mind doing it when needed.

Thanks for point! I can restart it manually next time

**Pull request merged**